File: frontend/src/api/annotate.ts
import { API_URL, ApiError, request } from "./http";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface UserItem {
  entry_id: string;
  author_channel_id: string;
  author_display_name: string;
  comment_count: number;
  my_annotated_count: number;
  my_pending_count: number;
}

export interface DatasetUsersResponse {
  dataset_id: string;
  dataset_name: string;
  total_users: number;
  total_comments: number;
  annotated_comments_by_me: number;
  items: UserItem[];
}

export interface MyAnnotation {
  label: string;
  justificativa: string | null;
  annotated_at: string;
}

export interface AnnotatorAnnotation {
  annotator_name: string;
  label: string;
  justificativa: string | null;
  annotated_at: string;
}

export interface CommentWithAnnotation {
  comment_db_id: string;
  text_original: string;
  like_count: number;
  reply_count: number;
  published_at: string;
  my_annotation: MyAnnotation | null;
  all_annotations?: AnnotatorAnnotation[] | null;
}

export interface UserCommentsResponse {
  entry_id: string;
  author_display_name: string;
  author_channel_id: string;
  comments: CommentWithAnnotation[];
}

export interface AnnotationResult {
  annotation_id: string;
  comment_db_id: string;
  label: string;
  conflict_created: boolean;
}

export interface DatasetProgress {
  dataset_id: string;
  dataset_name: string;
  total_comments: number;
  annotated: number;
  bots: number;
  humans: number;
  percent_complete: number;
}

export interface ImportResult {
  imported: number;
  updated: number;
  skipped: number;
  errors: string[];
}

export interface AnnotatorProgress {
  annotator_id: string;
  annotator_name: string;
  dataset_id: string;
  dataset_name: string;
  total_comments: number;
  annotated: number;
  bots: number;
  humans: number;
  percent_complete: number;
}

// ─── API ────────────────────────────────────────────────────────────────────

export const annotateApi = {
  listUsers: (datasetId: string, token: string) =>
    request<DatasetUsersResponse>(`/annotate/users?dataset_id=${datasetId}`, {}, token),

  getComments: (entryId: string, token: string) =>
    request<UserCommentsResponse>(`/annotate/comments/${entryId}`, {}, token),

  submit: (
    data: {
      comment_db_id: string;
      label: "bot" | "humano";
      justificativa?: string | null;
    },
    token: string
  ) =>
    request<AnnotationResult>("/annotate", { method: "POST", body: JSON.stringify(data) }, token),

  myProgress: (token: string) => request<DatasetProgress[]>("/annotate/my-progress", {}, token),

  allProgress: (token: string) => request<AnnotatorProgress[]>("/annotate/all-progress", {}, token),

  importAnnotations: async (
    data: {
      dataset_id?: string;
      dataset_name?: string;
      video_id?: string;
      annotations: Array<{
        comment_db_id: string;
        label: "bot" | "humano";
        justificativa?: string | null;
      }>;
    },
    token: string,
    onProgress?: (sent: number, total: number) => void
  ): Promise<ImportResult> => {
    const CHUNK_SIZE = 2000;
    const all = data.annotations;
    const total = all.length;
    const firstBatch = all.slice(0, CHUNK_SIZE);
    const hasMore = total > CHUNK_SIZE;

    const result = await request<ImportResult>(
      "/annotate/import",
      {
        method: "POST",
        body: JSON.stringify({
          ...data,
          annotations: firstBatch,
          done: !hasMore,
        }),
      },
      token
    );
    onProgress?.(firstBatch.length, total);

    if (!hasMore) return result;

    let offset = CHUNK_SIZE;
    let totalImported = result.imported;
    let totalUpdated = result.updated;

    while (offset < total) {
      const chunk = all.slice(offset, offset + CHUNK_SIZE);
      const isLast = offset + chunk.length >= total;
      const chunkResult = await request<{
        total_imported: number;
        total_updated: number;
        chunk_received: number;
        done: boolean;
      }>(
        "/annotate/import-chunk",
        {
          method: "POST",
          body: JSON.stringify({ annotations: chunk, done: isLast }),
        },
        token
      );
      totalImported += chunkResult.total_imported;
      totalUpdated += chunkResult.total_updated;
      offset += chunk.length;
      onProgress?.(offset, total);
    }

    return {
      imported: totalImported,
      updated: totalUpdated,
      skipped: result.skipped,
      errors: result.errors,
    };
  },

  downloadExport: async (
    format: "json" | "csv",
    token: string,
    datasetId?: string
  ): Promise<void> => {
    const qs = new URLSearchParams();
    qs.set("format", format);
    if (datasetId) qs.set("dataset_id", datasetId);

    const res = await fetch(`${API_URL}/annotate/export?${qs}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) {
      const body = (await res.json().catch(() => ({}))) as { detail?: string };
      throw new ApiError(body.detail ?? "Erro ao exportar anotações.", res.status);
    }
    const blob = await res.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `annotations.${format}`;
    a.click();
    URL.revokeObjectURL(url);
  },
};

File: backend/schemas/annotate.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

# ─── Request ─────────────────────────────────────────────────────────────────


class AnnotationCreate(BaseModel):
    comment_db_id: uuid.UUID
    label: Literal["bot", "humano"]
    justificativa: str | None = None

    @model_validator(mode="after")
    def justificativa_required_for_bot(self):
        if self.label == "bot" and not (self.justificativa or "").strip():
            raise ValueError("Justificativa é obrigatória para classificação 'bot'.")
        return self


class AnnotationImportItem(BaseModel):
    comment_db_id: uuid.UUID
    label: Literal["bot", "humano"]
    justificativa: str | None = None


class AnnotationImport(BaseModel):
    """Aceita o mesmo formato JSON exportado pelo export."""

    dataset_id: uuid.UUID | None = None
    dataset_name: str | None = None
    video_id: str | None = None
    annotations: list[AnnotationImportItem] = Field(min_length=1)
    done: bool = True


class AnnotationImportChunk(BaseModel):
    """Batch adicional de anotações para import paginado."""

    annotations: list[AnnotationImportItem] = Field(min_length=1)
    done: bool = False


class ImportChunkResponse(BaseModel):
    total_imported: int
    total_updated: int
    chunk_received: int
    done: bool


# ─── Response ────────────────────────────────────────────────────────────────


class AnnotationResult(BaseModel):
    annotation_id: uuid.UUID
    comment_db_id: uuid.UUID
    label: str
    conflict_created: bool


class MyAnnotation(BaseModel):
    label: str
    justificativa: str | None
    annotated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnnotatorAnnotation(BaseModel):
    annotator_name: str
    label: str
    justificativa: str | None
    annotated_at: datetime


class CommentWithAnnotation(BaseModel):
    comment_db_id: uuid.UUID
    text_original: str
    like_count: int
    reply_count: int
    published_at: datetime
    my_annotation: MyAnnotation | None
    all_annotations: list[AnnotatorAnnotation] | None = None


class UserCommentsResponse(BaseModel):
    entry_id: uuid.UUID
    author_display_name: str
    author_channel_id: str
    comments: list[CommentWithAnnotation]


class UserItem(BaseModel):
    entry_id: uuid.UUID
    author_channel_id: str
    author_display_name: str
    comment_count: int
    my_annotated_count: int
    my_pending_count: int


class DatasetUsersResponse(BaseModel):
    dataset_id: uuid.UUID
    dataset_name: str
    total_users: int
    total_comments: int
    annotated_comments_by_me: int
    page: int
    page_size: int
    total_pages: int
    items: list[UserItem]


class DatasetProgress(BaseModel):
    dataset_id: uuid.UUID
    dataset_name: str
    total_comments: int
    annotated: int
    bots: int
    humans: int
    percent_complete: float


class ImportResult(BaseModel):
    imported: int
    updated: int
    skipped: int
    errors: list[str]


class AnnotatorProgress(BaseModel):
    annotator_id: uuid.UUID
    annotator_name: str
    dataset_id: uuid.UUID
    dataset_name: str
    total_comments: int
    annotated: int
    bots: int
    humans: int
    percent_complete: float
